templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,261 @@
+.Parameter temp_puppetres
+  Location of the temporary puppetres.dll file.
+#>
+function Script:Reset-PuppetresDLL {
+  param(
+    [Parameter(Mandatory=$true)]
+    [String] $temp_puppetres
+  )
+  begin {
+    # _Never_ use the $InstallDir top-level parameter to try and find the InstallDir
+    # for the puppetres workaround. The workaround should _always_ fetch the InstallDir
+    # from the registry. This is so a user can specify a different InstallDir than the
+    # directory where the current package is installed and the workaround will still work
+    # for the already installed package
+    $InstallDir = Find-InstallDir
+    if (Test-Path $temp_puppetres -and -not (Test-Path "$InstallDir\puppet\bin\puppetres.dll")) {
+      Write-Log "Restoring puppetres.dll"
+      Move-Item -Path $temp_puppetres -Destination "$InstallDir\puppet\bin\puppetres.dll"
+    }
+  }
+}
+
+<#
+.Synopsis
+  Write message to location of $Logfile
+.Parameter message
+  String containing the message to write
+#>
+function Script:Write-Log {
+  param(
+    [Parameter(Mandatory=$true)]
+    [String] $message
+  )
+  begin {
+    $message | Out-File -FilePath $Logfile -Append
+  }
+}
+
+<#
+.Synopsis
+  Take control of the installation lockfile, fail if the lock
+  already exists
+.Parameter install_pid_lock
+  Location of the installation pid lock file
+#>
+function Script:Lock-Installation {
+  param(
+    [Parameter(Mandatory=$true)]
+    [String] $install_pid_lock
+  )
+  begin {
+    Write-Log "Locking installation"
+    if (Test-Path $install_pid_lock) {
+      Write-Log "Another process has control of $install_pid_lock! Cannot lock, exiting..."
+      throw
+    } else {
+      $PID | Out-File -NoClobber -FilePath $install_pid_lock
+    }
+    Write-Log "Locked"
+  }
+}
+
+<#
+.Synopsis
+  Release control of the installation lockfile
+.Parameter install_pid_lock
+  Location of the installation pid lock file
+#>
+function Script:Unlock-Installation {
+  param(
+    [Parameter(Mandatory=$true)]
+    [String] $install_pid_lock
+  )
+  begin {
+    Write-Log "Unlocking installation"
+    if (Test-Path $install_pid_lock) {
+      if ((Get-Content $install_pid_lock) -ne $PID) {
+        Write-Log "Another process has control of $install_pid_lock! Cannot unlock, exiting..."
+      } else {
+        try {
+          Remove-Item -Force $install_pid_lock | Out-Null
+          Write-Log "Unlocked"
+        } catch {
+          Write-Log $_
+        }
+      }
+    }
+  }
+}
+
+# ************** Execution start **************
+$ErrorActionPreference = "Stop"
+$service_names=@(
+  "puppet",
+  "pxp-agent",
+  "mcollective"
+)
+try {
+  Write-Log "Installation PID:$PID"
+  $install_pid_lock = Join-Path -Path (Split-Path -Parent (puppet.bat config print agent_catalog_run_lockfile)) -ChildPath 'puppet_install.pid'
+  Lock-Installation $install_pid_lock
+  # Wait for the puppet run to finish
+  #
+  # We *must* wait for Puppet Agent to finished applying its prior catalog

```

Comment:
Nit: Typo here -- "to finished applying".

Also, this comment should be expanded on a bit more. Probably something like

> We *must* wait for the puppet agent to finish applying its catalog before we stop all of our services. Otherwise if the catalog has additional resources that manage our services (e.g. such as those from the PE module), then the install will fail to proceed.

i.e. it isn't something specific to pxp-agent, it applies to all services.